### DISCUSSION TITLE
Question about the Lidar

### DISCUSSION THREAD

**Discussion by NDR008** [trackmania-rl/tmrl#29] [Q&A]
Hi everyone, I am working on an ambitious project (it is in particularly ambitious because I am not a hard core CS person, and am trying to do this project for my MSc in CS final project).

But it is similar to this project in that my project proposal looks at establishing tools/framework for interfacing a PlayStation 1 emulator with Python to then develop an agent that is developed using SAC to play Gran Turismo 1. So in part similar to what Sony did with Sophy (but less ambitious on how well I hope it will work).

Early on I was sure I can simply use openCV to process the screen and perform some edge detection to do something similar to your Lidar lines. However I've struggled with this approach. I then took a detour at using Ghidra to reverse engineer the game.
Made some useful progress in finding internal variables like vehicle speed - which will make things simpler and more similar to your solution for TrackMania 2020 (i.e. I do not need to use a NN to read the speed and so on).

But I could not make head-way on the lidar. Where can I look at how you've implement the spatial recognition / lidar?

Oh by the way - I read somewhere that you could only get 20FPS of screen capture?
Here I was experimenting with different ways: https://youtu.be/0t97yrOAKXs?t=1398
I was able to find a library called d3d that gave me VERY fast screen capture.
(Ironically for my project I do not plan to use it because I think I will use the emulator to share over the display-data directly to the python environment).

**Comment by yannbouteiller** [comment]
Hi!

The way we compute the LIDAR in the TM LIDAR environment is very specific to the plain road in TrackMania and probably won't work in Gran Turismo. We just look for black pixels on the trajectory of each ray, you can find the corresponding code [here](https://github.com/trackmania-rl/tmrl/blob/c73c6db60541cc325e6db428650ca11a7e6e07fd/tmrl/custom/utils/tools.py#L85)

This was just to speed up training in simple "toy" environments: nowadays we use full unprocessed images with CNNs.

The 20FPS value is not a bottleneck with screenshot capture. It is only an arbitrary choice for the control frequency that you can change in the `config.json` file if you wish. We chose 20FPS by default because (1) it leaves enough time between frames to infer the first dynamic moments of the car from the short history of 4 frames that we use by default, (2) it leaves enough time to compute the output of rather big deep learning models, and (3) it leaves enough time to capture screenshots.

**Reply by yannbouteiller** [reply]
(BTW, unless you pause your simulator to do your computations, the shortest path to your objective is to use tmrl with a custom pipeline, as described in the long tutorial - careful though: the current documentation is for tmrl 0.4.0, which is not yet released on PyPI, you need to clone it from GitHub atm)

**Reply by NDR008** [reply]
Thank you very much for replying!

I do not know if you meant to suggest or not using tmrl for my own project.
To be honest... I did not think of it, but worth thinking about - though it depends on how much of tmrl would I end-up not using.
Developing my own pipelines was anyway things I assumed I would have to do (and actually the part of this project taking the most time).
Coming up with a trajectory intersection with the track edge has been were I have been mostly stuck when it comes to the computer vision. Ironically the simplicity of a PS1 Gran Turismo graphics are making it harder to process the image with things like Canny Edge, because sometimes, the track color and the side barrier color have a lower contrast than say the car shadows. It seems very convenient that you could look for the black pixel instead...

But I am glad to see the method anyway, because that was also my intention - to do the ray-projection to track edges from the "1st person". Sony themselves used full perfect track position for their agent (since they had a special API access to in game parameters).

Regarding: (a) [BTW, unless you pause your simulator to do your computations] - (b) [the shortest path to your objective is to use tmrl with a custom pipeline, as described in the long tutorial]
Regarding (a) : I actually can easily pause the emulator (it is one of the controls I can exchange between the python environment and emulator) - but, why do you think I need to do that?
(b) Why is tmrl overcoming that barrier?

**Reply by yannbouteiller** [reply]
Because if you do not pause your simulator between time-steps for compuations, you will need to handle real-time, i.e., the fact that computations take random amounts of time, which is what rtgym is made to handle (tmrl is a framework for pratical robotics built around rtgym). If you approach this naively by just capturing frames as fast as you can and feeding them to a neural network in between, you will likely end up with a highly non-Markov environment that any RL approach will struggle to solve, and the performance of your final policy will be hardware-dependent. tmrl provides a skeletton for your pipeline that will accelerate your effort dealing with this kind of issues, especially since what you want to do basically amounts to just adapting the custom rtgym interfaces provided for TrackMania to GranTurismo.

**Reply by yannbouteiller** [reply]
(But it only makes sense if you are interested in developping a proper robot controller; if you want to develop a videogame-backend-like controller, what makes most sense is to pause the simulator between frames)

**Reply by NDR008** [reply]
So the naiveness I am still carrying is - I do not have a realistic sense of the time needed between each state-step.
Of course, since I am working with PS1 emulated hardware, game loops on the PS1, are handled normally no faster than 50Hz (for an EU/PAL game). Unlike PC games of course, the game runs at relatively speaking fixed speed (since it is standardised hardware). The emulator I am using, has a lua script interpreter which I am using to process in-game states and transmit over to my python environment. So some assumptions of my dissertation work was (so far) that I can a) get everything done within each frame (which would take 1/50s) or perform the data exchange at a lower frequency - example 25Hz.

However, if I understood your correctly, tmrl is able to cope with the variability of frequency in order to work in a real-time environment?
It is getting more interesting. I think I will start of by cloning this git and see what it does and how it works as is.
(So far I have only looked at youtube videos of it).

**Reply by yannbouteiller** [reply]
More precisely, it is `rtgym` that uses some thread and delay trickery to create a proper fixed-frequency Gym environment from any real-time application. In a real-time application, by definition, you cannot stop time before computing and applying your actions, or before retrieving your next observations.

(PS: if you are going to try out the pre-trained policies, try the LIDAR version, because the image version has been broken by a change in the last version of TrackMania apparently, I am trying to figure out why)

**Reply by NDR008** [reply]
Thanks, been cleaning up my Conda environments, but will try to spend some time this weekend to get it working. :)

**Reply by NDR008** [reply]
So I am trying to manually install tmrl.
Some questions & issues...

1) Is TMRL_FOLDER = HOME_FOLDER / "TmrlData" where the python RL related data will go into? (I ask because I intend to not use the .home() path)

2) [OpenPlanet should also have created a folder named OpenplanetNext] (from the tutorial)
Actually OpenPlanet created a folder named Openplanet in the folder:
```C:\Program Files\Epic Games\TrackmaniaNext\Openplanet```
Inside Openplanet is ```Fonts, Plugins, Scripts, cacert.pem```
I am guessing this is the folder intended and not OpenplanetNext right?

**Reply by yannbouteiller** [reply]
I suppose it is, I have not installed TrackMania from Epic Games and for me the OpenPlanet folder is in the home directory so I don't know, but I guess it is the right place if it has the Scripts folder.

You are right that TMRL_FOLDER = HOME_FOLDER / "TmrlData" is where the TmrlData folder is, if you want to change this you have to do it in the source code because I am not aware of a way to give a custom destination folder to pip (and the TmrlData folder is super useful so we chose the home folder for it)

**Reply by NDR008** [reply]
OK, so I have manually gone through the whole process of installing, and went to try copying:
```Load the tmrl-test track into your TrackMania game:```
and am stuck and am starting to wonder if I have the correct TrackMania installed or in the correct place....

When II go to trackmania.com as per your instructions I find just a webnsite with "Play For Free" button, which takes me the the ubi-store, and launches my Ubi-Soft download client...
I cannot see a traditional way of installing TrackMania...

So I actually cannot find a maps folder... major bummer. Also seems like I need to buy the club edition of the game according to the OpenPlanet readme.

**Reply by yannbouteiller** [reply]
Oh but Ubisoft Connect IS the traditional way of installing TrackMania. I just thought you had installed it via Epic Games because your OpenPlanet folder is weirdly in the Epic Games directory. You have to put the `tmrl-test.Map.Gbx` file (from `TmrlData/resources`) in the custom map folder, you can find this folder in-game in create>track editor>edit a track>my local tracks (there you should find an icon that enables you to open your custom track folder)

**Reply by yannbouteiller** [reply]
You should not need the club edition, I had the OpenPlanet script signed by the developer of OpenPlanet so that it can be used with the free edition.

**Reply by NDR008** [reply]
Oh.... getting there, so it seems that the folder is in C:\Users\USERNAME\Documents\Trackmania\Maps
The epic games... understood, I must have installed it via EpicGames previously. Following the Ubi-Soft client just points out that the game is already installed.

Anyway loaded the test track. :) Let me see what else I got to do.

**Reply by NDR008** [reply]
OK, so, I am stuck at....
Hitting "3" does not take me to the cockpit view.
(I got there anyway by other ways).
But I am "in" the cockpit unable to hide the car...

Here is as far as I got :(
https://imgur.com/a/W9wGQiM

*SOLVED* https://imgur.com/2LW2iJ7

**Reply by NDR008** [reply]
Is this how the provided agent should behave?
https://youtu.be/ELgNSKIIH_A

**Reply by yannbouteiller** [reply]
No, this looks like the broken CNN agent (not the LIDAR agent). It looks like the `config.json` file that you are editing is not the one used by TMRL

**Reply by yannbouteiller** [reply]
You can see this because the window is resized to the very small size, which is used by the image/CNN environment. The LIDAR/MLP environment should resize your TrackMania window to a much larger size

**Reply by yannbouteiller** [reply]
Unless you edited the code, the `config.json` used by `tmrl` is in the `TmrlData` folder of your home directory

**Reply by yannbouteiller** [reply]
(And BTW if you cloned and modded the repo, you want to install the library by doing `pip install -e .` where your `setup.py` file is located. If you do `pip install tmrl` instead, this will install the old 0.3 version that is on PyPI)

**Reply by NDR008** [reply]
Need to figure out where/what is tmrl pointing to then, because it is th TMRLDATA folder that I modified for the setup.py... hmmmm
but things can go messy with mixes of pip / conda / manual-setups....

**Reply by NDR008** [reply]
I did not know this:
pip install -e .

I cloned the git, and then tried to work my way through setup.py.... HAH! 👍 
This is cool, IF I can get this to work on GT1, I wonder if I can say I did enough for my dissertation.
Anyway, thanks.
By the way - is it ok if I keep using this discussion tab or would you rather have my ping you elsewhere? Especially when I step into the GT stuff?

**Reply by yannbouteiller** [reply]
You can keep using this discussion, np, it is here for that :)

**Comment by yannbouteiller** [comment]
(BTW don't use the image/CNN agent yet because I am training a new one currently, using a different preprocessor. The up-to-date code for that is on the `yb/dev` branch, I'll push to `master` shortly)

**Reply by NDR008** [reply]
ah, figured, evn tho I had manually modified things... actually I still have a TmrlData folder under user
I wonder were / what I need to modify to mak use of my other folder.

**Reply by NDR008** [reply]
Anyway, went for the default TmrlData folder now and woohoo it works! 👍 
Thanks!!!

**Reply by NDR008** [reply]
Are these warnings:
```
C:\Users\USER\anaconda3\envs\tmrl\lib\site-packages\rtgym-0.7-py3.10.egg\rtgym\envs\real_time_env.py:365: UserWarning: Time-step timed out. Elapsed since last time-step: 1.5455756187438965
C:\Users\USER\anaconda3\envs\tmrl\lib\site-packages\rtgym-0.7-py3.10.egg\rtgym\envs\real_time_env.py:365: UserWarning: Time-step timed out. Elapsed since last time-step: 0.33658623695373535
C:\Users\USER\anaconda3\envs\tmrl\lib\site-packages\rtgym-0.7-py3.10.egg\rtgym\envs\real_time_env.py:365: UserWarning: Time-step timed out. Elapsed since last time-step: 0.3390779495239258
```
Part of the magic that rtgym is working on, and some sort of "too unsynced" situation?

**Comment by NDR008** [comment]
https://youtu.be/oOU20m6e62U
By the way, is there a reason that the lap time is so non-repeatable?
Is it because of the time-stamp / real-time problem variance or something on my end or something totally different?

**Reply by yannbouteiller** [reply]
That looks more like it :)

The lap time is non-repeatable because both the environment and SAC policies are by nature very stochastic.

The environment is stochastic because of its real-time nature: contrary to a video-game back-end controller that would control the car between every frame in a deterministic fashion, `rtgym` samples at an elastic frequency of 20 FPS in this case, which means that it does its best for your control frequency to be 20 FPS *on average*. In practice, there is no way to ensure that 2 frames are precisely spaced of, e.g, 0.05s, because you would need dedicated real-time hardware for this (or to pause your simulator at precisely chosen points in time, but `tmrl` is geared for real-time, not for this).

SAC policies are stochastic in nature because SAC literally maximizes their stochasticity (entropy).

If you get many "time step timed out" warnings, it also means that your PC is not powerful enough or too busy with other tasks to run the MLP at 20Hz, which harms the performance a bit.

**Reply by NDR008** [reply]
I do not get it frequently enough, but my PC is very capable... 16 thread capability at 4GHz with 32GB at 3600MHz and a 4GB GTX1650, most machine learning tasks ran faster locally than on Google Colab.
Does the run inference on GPU work? (will try it anyway).
Ermmm fail: ```TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.```
I guess many places a numpy array was used. During one my last machine learning assignments, I made a great deal of effort to avoid numpy and stick to native tensors, that is how I could gain a lot of performance boost.

Coming back to this question, so if I were able to pause emulation and/or have enough time to computer everything in between frames, I would not suffer from this issue and potentially be fine?

I do not know if you read the Sony paper on Sophy AI:
https://www.nature.com/articles/s41586-021-04357-7
Their work is amazing because they got their AI to have reward/penalty functions even for sportsmanship, and could perform extremely well on track against human players.

My goal is not that high, I am targeting time-trial, so very much in the same level achieve by your work.
But the pipeline is going to be a bitch... in particular the detecting the track edge.

The many shades of grey are tricky to use CV:
![image](https://user-images.githubusercontent.com/34957530/211172416-d6036a6a-7d3f-4275-b384-e3f0388bb998.png)

The wrong detail is captured no matter what I tried so far:
![image](https://user-images.githubusercontent.com/34957530/211172450-e4d07b92-8d09-443d-a045-23b6c77d8277.png)

The other thing I am thinking of doing is trying to detect the centre of the track from the discontinuous white markings.

**Reply by yannbouteiller** [reply]
Oh, you found a bug haha, I'll fix that, I haven't tested cuda inference for a long time (that being said, it is typically slower than cpu inference; cuda is usually good for training on batches, not for processing a single sample at inference time)

**Reply by yannbouteiller** [reply]
Okay the cuda bug should be fixed now

**Reply by yannbouteiller** [reply]
For your task, I would personally not even try to use traditional CV. I would do character detection to retrieve the speed in the bottom right corner, and go for a CNN directly. How do you plan to compute your reward function? You can analyze the position of the red dot on the map in the bottom left corner, and give a positive reward each time it moves in the good direction (like, consider the map is an ensemble of equally-spaced checkpoints, and give the same reward for each passed checkpoint - this is basically how the reward function works in TrackMania BTW).

**Reply by yannbouteiller** [reply]
"Coming back to this question, so if I were able to pause emulation and/or have enough time to computer everything in between frames, I would not suffer from this issue and potentially be fine?" > Yes, but your approach would also be wrong for real-world self-driving, which would be a shame in my opinion given that this is a car simulation. But it is also how you can beat human players and get a short paper accepted in Nature ;)

**Reply by yannbouteiller** [reply]
(More seriously, the most important for this is that you are mindful of the Markov property. Real-time considerations are a detail, what is very important for you is that you ensure your model can infer important dynamics such as linear and angular velocity/acceleration)

**Reply by NDR008** [reply]
Sony did not pause, but they had insane computation power, and all in-game data pumped out by the game in an idealised manner.

For speed retrieval basic vehicle state parameters - that is mostly easy / done - I have figured out the memory addresses of several parameters such as gear, speed, turbo-boost, etc.
https://github.com/NDR008/TensorFlowPSX/blob/8abda279a98a4cd4ccff995cc54865304205c9e4/ReduxLua/gt.lua#L182
I can even remove the whole HUD by flipping 1 flag of the in-game emulated memory, to give any CNN a pure "virtual world" image.

When you say go for CNN - still something like the LIDAR would still be necessary right? I do not have the "black pixels" to help me out.

Regarding reward - that dilemma has been on my mind, and I saw that you wrote about it too, that speed would be a naïve reward.
There is a variable I found which seems to be related to "track segment" but... the variable I found suggests segments are pretty big (like start-finish line to first corner is one segment) - so the agent would not be able to determine if it did well or not.
The map was also on my mind, but I was not sure how to encode it as I would have to employ a CNN to visually understand it OR spend some more time trying to reverse engineer the map function in Ghidra (if I can find the related function - C and reverse engineering is also part of the "new" challenge stuff). 

If I choose to pause or skip frames, I could investigate the ability of an AI agent at different frequency of observations. It is one of the things I am critical about on Sony publication.

**Reply by NDR008** [reply]
On the bugfix, I should have persevered. I had tried cpu().a.numpy() - misunderstood the syntax of that call.

**Reply by yannbouteiller** [reply]
A CNN enables you to use raw images, no need for a LIDAR. The LIDAR is just faster to train because it uses a simple MLP, but with the right hyperparameters CNNs are still a realistic solution (with the new hyperparameters that I pushed in the default config.json of Release 0.4.0, `tmrl` seems to train CNNs quite quickly in TrackMania, I am impressed :P)

For your reward function, the way I look at it, reverse-engineering of this map seems the simplest solution. Of course you could use the speed which would be much easier, but then your AI would only learn to go fast.

**Reply by NDR008** [reply]
Or to do burnouts and get the speedo-high without going anywhere.
Interesting, the full-on CNN vs LIDAR, if workable would be... really nice, as I could relate it to the real auto-industry dilemma - the Tesla non Lidar vs the everybody else with RADAR + LIDAR + Cameras...

Thanks I guess I will try spending some time on encoding the map, as for the CNN, I will probably pester you a bit on that later on.
Did you use a custom built CNN or did you use any pre-trained models like ResNet?

**Reply by yannbouteiller** [reply]
I use the custom and very simple CNN that you can find [here](https://github.com/trackmania-rl/tmrl/blob/master/tmrl/custom/custom_models.py#L475)

We have experimented with a few more complex models (pre-trained or not) such as ResNet, EffNet, MobileNet etc, but without much success so far. Truth is, deep RL does not like fancy tricks that supervised classification loves such as batch normalization. Plus, real-time keeps you from using anything too computational if you cannot run it fast enough.

**Comment by NDR008** [comment]
By the way - is the CNN version (non-LIDAR) working now?

**Reply by yannbouteiller** [reply]
It is working, and starting today it is the default on PyPI. I have just released `0.4.0`

I am still training the CNN baseline and I will silently update the release over time, but the current baseline is sensible (at least when using the canadian skin, I have not tried with other skins)

**Reply by yannbouteiller** [reply]
Also, interestingly, it kinda generalizes to the end of the `tmrl-train` track although it has never seen it during training, which I expected would be hard for a CNN.

**Reply by NDR008** [reply]
Not sure what I am doing wrong...
I reinstalled tmrl, redownloaded the resource file from github, overwrote everything that was inside the trmldata folder, and I know have an agent that drives continuously left?

(Swapping back to lidar works)

**Reply by NDR008** [reply]
Nevermind, realised this is trained with a visible car
Is there a reason to making this on your end?

**Reply by yannbouteiller** [reply]
Yes, the CNN policy is trained on the default camera using the canadian flag skin.

The reason for doing this is that it makes the environment more Markov than giving only the front camera. In particular the agent can know whether it is drifting by looking at the marks behind the car, which is not possible with the front camera only (front camera only would likely require an RNN for achieving this)

**Comment by NDR008** [comment]
Since this discussion was about the Lidar, I think it would be better to close this discussion and open a new, but my next question is about training.
I want to understand a bit more about tmrl, because figuring out how to adapt for my needs.

The instructions say:
Record a reward for this track:
Execute:
Press e to start recording
Complete the track

I am not understanding what this means/does. Is the reward not a reward function that you coded?
Is it something that you learn by playing the game yourself?

Also I keep getting this warning from teh trainer.
INFO:root:--- NOW RUNNING SAC on TrackMania ---
wandb: Currently logged in as: tmrl. Use `wandb login --relogin` to force relogin
wandb: ERROR Error while calling W&B API: run tmrl/tmrl/SAC_4_imgs_pretrained was previously created and deleted; try a new run name (<Response [409]>)
wandb: ERROR Error while calling W&B API: run tmrl/tmrl/SAC_4_imgs_pretrained was previously created and deleted; try a new run name (<Response [409]>)

**Reply by yannbouteiller** [reply]
Hi! So, for the warning, you want to change the name of your run in `config.json` (or use your own `wandb` credentials - you get the warning because you try to log learning curves on `wandb` using a run name that has been used already)

About the reward function, the code that computes the reward in `tmrl` uses a trajectory that you need to record by following the instructions that you cite. The way this trajectory is used is described [here](https://github.com/trackmania-rl/tmrl#a-clever-reward)
